### PR TITLE
Include version info in Installer

### PR DESCRIPTION
**Pull request by SinghRajenM**
As of now, there is no version info available on the detail section of installer properties.
It is good to have version info.

file_info

### CODE REVIEW

File: PowerEditor/installer/nsisInclude/gobalDef.nsh [hunk 1]
```diff
@@ -34,6 +34,12 @@
 !define VERSION_MAJOR 7
 !define VERSION_MINOR 0
 
+!define CompanyName		"Don HO don.h@free.fr"
+!define Description		"Notepad++ : a free (GNU) source code editor"
+!define Version 		"7.0.0.0"
```
Comment (justanotheranonymoususer) [inline PowerEditor/installer/nsisInclude/gobalDef.nsh]: This is yet another place to modify when a new version comes out. Why not use the macros above?

Comment (SinghRajenM) [inline PowerEditor/installer/nsisInclude/gobalDef.nsh]: If you are talking about existing macro APPVERSION, it can't be used directly as VIProductVersion expect version string X.X.X.X. We have to write many conditions to utilize APPVERSION. 
Reason is:
In V7 APPVERSION = 7
In V6.9.2 APPVERSION = 6.9.2
In V6.9 APPVERSION = 6.9

Comment (MAPJe71) [inline PowerEditor/installer/nsisInclude/gobalDef.nsh]: I think he ment using something like ...
``` NSIS
!define Version 		VERSION_MAJOR "." VERSION_MINOR
```
instead of ...
``` NSIS
!define Version 		"7.0.0.0"
```

Comment (SinghRajenM) [inline PowerEditor/installer/nsisInclude/gobalDef.nsh]: However, I was thinking why can't we remove macro APPVERSION (VERSION_MAJOR and VERSION_MINOR as well). Instead of changing its value for every release, just read the version info from notepad++.exe and use it.

This requires an additional nsis file (get version of npp exe and write version info in normal text file in form of macro with required format.) and its output (exe) file.
Later include call this exe in beginning of nsi script to generate text file and use these defined macro.
We have used this approach in many project. This will reduce the manual effort of writing version info everytime.

Comment (SinghRajenM) [inline PowerEditor/installer/nsisInclude/gobalDef.nsh]: @MAPJe71 Initially, I though to utilizing existing macros, but when I saw the history of these macros, I could not use these without writing conditions as the value. Consider V6.9.2, then major is 6, minor is 92. So break 92 as 9.2 and append .0 (6.9.2.0) while in V7 major is 7, minor is 0, so append .0.0 (so 7.0.0.0).

File: PowerEditor/installer/nsisInclude/GetVersion.nsi [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+SetCompressor /SOLID lzma 		; Compression technique
+RequestExecutionLevel user 		; This is to make output exe run as invoker (Admin privileges not required)
+!define EXEPATH "..\bin\Notepad++.exe"	; Absolute path
+
+
+Section "Main"
+	SetAutoClose true	; it simply writes version info in the text. So close immediatly after writing version info
+
+    DetailPrint "Getting version from ${EXEPATH}"
+    GetDllVersion "${EXEPATH}" $R0 $R1 	; Two values were read during compilation
```
Comment (MAPJe71) [inline PowerEditor/installer/nsisInclude/GetVersion.nsi]: The trailing comment doesn't make sense to me ...
- which two values?
- during compilation of what?

File: PowerEditor/installer/nsisInclude/GetVersion.nsi [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+	
+	;${GetFileVersion} "${EXEPATH}" $R0
+	;DetailPrint "Another way to get version: $R0"
```
Comment (MAPJe71) [inline PowerEditor/installer/nsisInclude/GetVersion.nsi]: Why another way?
Pick one and don't mention other way(s) _or_ mention the other way(s) and comment/explain why it's not the preferred way. Don't create confusion!

File: PowerEditor/installer/nsisInclude/GetVersion.nsi [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+	; if notepad++.exe version = 7.0.0.0
+	; 	$1 = 7.0.0.0 (File Version)
+	;	$2 = 7		 (Product Version)
+	;	$3 = 0		 (Minor Version)
+	
+	; if notepad++.exe version = 7.1.2.0
+	; 	$1 = 7.0.0.0 (File Version)
+	;	$2 = 7.1.2 	 (Product Version)
+	;	$3 = 12		 (Minor Version)
+
```
Comment (MAPJe71) [inline PowerEditor/installer/nsisInclude/GetVersion.nsi]: I think it's wise to get confirmation about the meaning of the different parts of the version number first i.e. before issuing a PR.

File: PowerEditor/installer/nsisInclude/GetVersion.nsi [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+			${Endif}
+		${Else}
+			StrCpy $2 "$R2.$R3.$R4"
+			StrCpy $3 "$R3$R4"
+		${Endif}
+	${Else}
+		StrCpy $2 "$R2.$R3.$R4.$R5"
+		StrCpy $3 "$R3$R4$R5"
+	${Endif}
+
```
Comment (MAPJe71) [inline PowerEditor/installer/nsisInclude/GetVersion.nsi]: Would be nice if it's possible to give the variables explanatory names.

**Comment by SinghRajenM** [comment]
Yes. It is generated from GetVersion.nsi but not during the build process.
This is one time activity and no need to again n again unless there is a
bug or some changes are required.

File: PowerEditor/installer/nsisInclude/GetVersion.nsi (hunk 2 above)
Comment (SinghRajenM) [inline PowerEditor/installer/nsisInclude/GetVersion.nsi]: Two values $ R0 and $ R1 which will be extracted from npp.exe.

File: PowerEditor/installer/nsisInclude/GetVersion.nsi (hunk 3 above)
Comment (SinghRajenM) [inline PowerEditor/installer/nsisInclude/GetVersion.nsi]: I don't think commented code creates confusion. If it really bothers people I'll remove commented code.

File: PowerEditor/installer/nsisInclude/GetVersion.nsi (hunk 5 above)
Comment (SinghRajenM) [inline PowerEditor/installer/nsisInclude/GetVersion.nsi]: I use these register variables because
: No declaration needed (var blah).
: Less LOC
: No name conflict (not applicable here very less code)
: Bit faster than normal variable

File: PowerEditor/installer/nsisInclude/GetVersion.nsi (hunk 2 above)
Comment (MAPJe71) [inline PowerEditor/installer/nsisInclude/GetVersion.nsi]: I get that the two $Rx variables/registers will contain values that are read from Notepad++ (i.e. ${EXEPATH}) but it's not clear what's the meaning of these values:
- ProductVersion and FileVersion?
- major and minor version?
- ...

File: PowerEditor/installer/nsisInclude/GetVersion.nsi (hunk 4 above)
Comment (MAPJe71) [inline PowerEditor/installer/nsisInclude/GetVersion.nsi]: Correct about APPVERSION but AFAIK FileVersion and ProductVersion are always the same for Notepad++.

File: PowerEditor/installer/nsisInclude/GetVersion.nsi (hunk 3 above)
Comment (MAPJe71) [inline PowerEditor/installer/nsisInclude/GetVersion.nsi]: Why did you add it?

File: PowerEditor/installer/nsisInclude/GetVersion.nsi (hunk 5 above)
Comment (MAPJe71) [inline PowerEditor/installer/nsisInclude/GetVersion.nsi]: I don't think it's a time critical part and [self explanatory code](https://github.com/notepad-plus-plus/notepad-plus-plus/blob/master/CONTRIBUTING.md#always-prefer-a-variable-name-that-describes-what-the-variable-is-used-for) is preferred (see Style Guide)

**Comment by MAPJe71** [comment]
IMO binaries in a repository are security liability but sometimes you can't get around them e.g. 3rd party libraries/tools. In this case the source is part of the project so I'd say build it during the build process but I'll leave that decision to @donho.

File: PowerEditor/installer/nsisInclude/GetVersion.nsi (hunk 2 above)
Comment (SinghRajenM) [inline PowerEditor/installer/nsisInclude/GetVersion.nsi]: I think below screenshot will answer your queries.

versioninfo

File: PowerEditor/installer/nsisInclude/GetVersion.nsi (hunk 4 above)
Comment (SinghRajenM) [inline PowerEditor/installer/nsisInclude/GetVersion.nsi]: No. They are not same. Refer the screenshot (attached above) of actual notepad++.exe. You will find the difference.
I think that make sense.

File: PowerEditor/installer/nsisInclude/GetVersion.nsi (hunk 5 above)
Comment (SinghRajenM) [inline PowerEditor/installer/nsisInclude/GetVersion.nsi]: This is way most of the nsis scripts are written. I feel these were introduce to simplify the task, otherwise there was no need to introduce registers $0~$9 and $R0~$R9.

**Comment by SinghRajenM** [comment]
> In this case the source is part of the project so I'd say build it during the build process

I feel it is not that much easy, moreover I don't how appveyor is configured. But if I take an example from Jenkins, it requires two jenkin jobs.
1. Complie GetVersion.nsi which creates GetVersion.exe. 
2. Actuall npp build job

In this case, job 1 will be slave job of job 2. So first job 2 should trigger job 1. Once job 1 is completed, move the artifact (GetVersion.exe) from job 1 to job2  and execute it which will write version info in form of macros.

File: PowerEditor/installer/nsisInclude/GetVersion.nsi (hunk 3 above)
Comment (MAPJe71) [inline PowerEditor/installer/nsisInclude/GetVersion.nsi]: As I said before pick one method/solution to be included in the official repository through PR.
Keep tries to your local development environment.

**Comment by MAPJe71** [comment]
Notepad++ build already has two dependencies, the Boost and Scintilla libraries, so it shouldn't be a problem for the AppVeyor build environment.
Furthermore, I don't see a procedural difference in building an executable from NSIS script source or C++ source.

**Comment by SinghRajenM** [comment]
> Notepad++ build already has two dependencies, the Boost and Scintilla libraries, so it shouldn't be a problem for the AppVeyor build environment.

If so, then GetVersion.nsi should be built in different job and artifact can be moved to "PowerEditor\installer\nsisInclude" before starting npp setup creation build.

Sorry, I don't how to create jobs and dependencies in AppVeyor as we have dedicated build engineers.

File: PowerEditor/installer/nsisInclude/GetVersion.nsi (hunk 4 above)
Comment (MAPJe71) [inline PowerEditor/installer/nsisInclude/GetVersion.nsi]: Product version and File version **are** the same (see [Notepad++ resource file](https://github.com/notepad-plus-plus/notepad-plus-plus/blob/master/PowerEditor/src/Notepad_plus.rc)).
But there's a difference in notation for the [`VERSION_DIGITALVALUE`](https://github.com/notepad-plus-plus/notepad-plus-plus/blob/master/PowerEditor/src/resource.h#L35) and the [`VERSION_VALUE`](https://github.com/notepad-plus-plus/notepad-plus-plus/blob/master/PowerEditor/src/resource.h#L34).
The property list of the Notepad++ v6.9.2.0 executable will show the digital value (i.e. comma-separated-digits) as dot-separated-digits (e.g. `6.9.2.0`) and the textual value will be visible as defined in the [resource file](https://github.com/notepad-plus-plus/notepad-plus-plus/blob/master/PowerEditor/src/resource.h#L34) i.e. `6.92`.

Comment (SinghRajenM) [inline PowerEditor/installer/nsisInclude/GetVersion.nsi]: Need to investigate if it is possible to extract from resource.h. 

URL: https://msdn.microsoft.com/en-us/library/windows/desktop/aa381058(v=vs.85).aspx might be helpful to understand versioning.
versioninfo2

Comment (MAPJe71) [inline PowerEditor/installer/nsisInclude/GetVersion.nsi]: What Windows version is that?
Strange that it shows the digital/fixed value for File version and textual/variable for Product version.

Yep, I've used automatic generated resource info in projects based on the linked MS info.
Notepad++ version numbering is not the same as MS tends to do it.
Here are some different methods I've seen over the years:
```
- Windows uses       : major.minor[.build[.revision]]
- .NET    uses       : major.minor[.revision[.build]]
- other used formats : major.minor[.maintenance[.build]]
                       major.minor[.build[.((majorRevision && 0x00FF)<<8)||(minorRevision && 0x00FF)]]
                       major.minor[.patch[.stage]]
```
More recent I've encountered projects mentioning they use [Semantic Versioning](http://semver.org/).

**Comment by cmeriaux** [comment]
About AppVeyor, the configuration is here : notepad-plus-plus/appveyor.yml  [link](https://github.com/notepad-plus-plus/notepad-plus-plus/blob/master/appveyor.yml)
There is one job ( see build_script section) that is executed with a matrix of parameters ( **platform** and **configuration** )
The job execute visual studio with this command :  

> msbuild notepadPlus.vcxproj /p:configuration="%configuration%" /p:platform="%platform%"

You can execute before this command the compliation of GetVersion.nsi then use your getVersion program to do what you want to do.

However, appVeyor is not used to build the notepad++ installer. It can't due to code certification.
appVeyor doesn't compile with boost support.

I hope it helps you

**Comment by cmeriaux** [comment]
I'd like also to have the git version to identify each version available (in case of people uses a nigthly build)
Git offer a command to do that :  [git describe](https://git-scm.com/docs/git-describe)

```
[21:42:41] christophe@Christophe-PC /cygdrive/d/dev/notepad-plus-plus  master
: git describe --tags --long
v7-36-g4fec265

```
=> My  repository HEAD is on version 7, 36 commits since tags, sha1 = 4fec265

**Comment by donho** [comment]
It's a small "Nice to have" feature. However, it's not worth to me to integrate all the codes for adding this small feature.

**Comment by SinghRajenM** [comment]
No problem!!!
